lib/AstOfLlbc.ml

```diff
@@ -348,26 +409,75 @@ let mk_deep_copy (e: K.expr) (l: K.expr) =
 
 (* Environment: expressions *)
 
+let is_var v2 v1 =
+  match v2 with
+  | Var (v2, _) -> v2 = v1
+  | _ -> false
+
+let assert_var = function
+  | Var (v2, ty) -> v2, ty
+  | _ -> assert false
+
+let assert_trait_clause_method = function
+  | TraitClauseMethod (clause_id, item_name, signature) -> clause_id, item_name, signature
+  | _ -> assert false
+
+(* Regular binders *)
+
 let lookup env v1 =
-  let i, (_, t, _) = findi (fun (v2, _, _) -> v1 = v2) env.binders in
+  let i, (_, t) = findi (fun (v2, _) -> is_var v2 v1) env.binders in
   i, t
 
 let lookup_with_original_type env v1 =
-  let i, (_, t, ty) = findi (fun (v2, _, _) -> v1 = v2) env.binders in
+  let i, (v, t) = findi (fun (v2, _) -> is_var v2 v1) env.binders in
+  let _, ty = assert_var v in
   i, t, ty
 
+(* Const generic binders *)
+
 let push_cg_binder env (t: C.const_generic_var) =
-  { env with cg_binders = (t.index, typ_of_literal_ty env t.ty) :: env.cg_binders }
+  { env with
+    cg_binders = (t.index, typ_of_literal_ty env t.ty) :: env.cg_binders;
+    binders = (ConstGenericVar t.index, typ_of_literal_ty env t.ty) :: env.binders;
+  }
 
 let push_cg_binders env (ts: C.const_generic_var list) =
   List.fold_left push_cg_binder env ts
 
 let push_binder env (t: C.var) =
-  { env with binders = (t.index, typ_of_ty env t.var_ty, t.var_ty) :: env.binders }
+  { env with binders = (Var (t.index, t.var_ty), typ_of_ty env t.var_ty) :: env.binders }
 
 let push_binders env (ts: C.var list) =
   List.fold_left push_binder env ts
 
+(* Clause binders, which only appear as function parameters, and hold an unknown
+   trait method (dictionary-style). *)
+
+type clause_binder = {
+  clause_id: C.trait_clause_id;
+  item_name: string;
+  pretty_name: string;
+  sig_info: sig_info;
+  t: K.typ;
+}
+
+let push_clause_binder env { clause_id; item_name; t; sig_info; _ } =
+  { env with
+    (* Perhaps (?) important for diff calculations *)
+    cg_binders = (C.ConstGenericVarId.of_int max_int, t) :: env.cg_binders;
```

Comment:
I don't understand this

Reply:
This is gone. It looks like I didn't understand this either.